### ISSUE TITLE
calicoctl expects specific name for calico/node container

### ISSUE THREAD

**Issue by caseydavenport** [projectcalico/calico-containers#1086]
When running under Kubernetes, the `calico/node` container name is unique.

e.g

```
k8s_calico-node.215a1ad0_calico-n7l79_default_6e29a86d-5868-11e6-9250-080027397496_d3d6d6b6
```

`calicoctl` expects the container name to be `calico-node`, and `calicoctl status` fails with the following message:

```
core@k8s-master ~ $ sudo ./calicoctl status
calico-node container not running
```

`calicoctl` should be more flexible here.

**Comment by robbrockbank** [comment]
I think we have a few options here:
-  Check for calico/node image name
-  Check for calico-node in the container name
-  Pass in the image name or container name into `calicoctl status`

The latter would be more generic but arguably harder to use.

@caseydavenport - do you have a particular preference for k8s?

**Comment by caseydavenport** [comment]
I think passing the container name into `calicoctl status` is a no-go.  The container name can sometimes be _really_ ugly to work with (see OP).

I think checking for `calico-node` in the name would be sufficient.  This is probably nicer anyway in case people want to use their own tagged version of calico-node (which would make the images not match).  e.g `caseydavenport/node`.

**Comment by gunjan5** [comment]
Fixed in the Go version # [115](https://github.com/projectcalico/libcalico-go/pull/115)

**Comment by JohnOmernik** [comment]
Hey all, I'd like to reopen this in that this solution only addresses K8s and not other uses (like DCOS/Mesos/Marathon)  Relying on the name of the container is not a good way, and shouldn't be used because there are cases we don't have control over the name (mesos doesn't include calico-node). 

One of the other options above "Check for calico/node image name" is better, however, this assumes that everyone who would run this would use the image from the internet (i.e. calico/node). I feel  some folks  (me included) would want to bring this image locally to a local docker repo, and it may be renamed. So once again, relying on the name is problematic. 

One option may be to use ENV or a parameter to pass in an "image string"   
Example: let's say instead of running calico/node, I run an image called myregistry.marathon.mesos:5005/calico-node and that's the image my nodes utilize for their operations.  If I could pass in an image string of calico-node (either as a parameter or a ENV variable) then calicoctl could check the running containers looking for a running container using the image that matches the image string I provided.  By default, calico could use calico/node that would work for all the current use cases (and K8s) and then for folks who run their own tagged image, we just have to specify our own image string. 


Using the image seems to be better than using the container name due to all the various things that can actually run containers and what they demand the image name be. 

Thoughts?

**Comment by caseydavenport** [comment]
@JohnOmernik thanks for chiming in - we definitely want to make sure this fix is generally applicable. 

I think an ENV variable + checking for the image rather than the name could be a valid solution.  

It looks like @gunjan5 has already done some work to make this command run even when Calico is not running in a container (https://github.com/projectcalico/libcalico-go/pull/149)

@gunjan5 I haven't looked at that fix - will that work for the containerized case without relying on the container name / image?

**Comment by gunjan5** [comment]
@caseydavenport in Go version, it just looks for the calico processes so container name wouldn't matter anymore

**Comment by kovert** [comment]
What about going through a list of things and going with the best match?   An environment variable  always wins, followed by a config file, the container named calico-node, the container with the label io.kubernetes.container.name=calico-node, the equivalent for mesos, the same for other things, all else fails go with the image name?  It seems like none of these are perfect, but making a best guess and providing a way for users to both override the best guess and set a system global when they know how to deploy it would make this most usable.

**Comment by n-marton** [comment]
Or you can use this version: https://github.com/n-marton/calico-containers/releases/tag/v0.22.0k8s  

--runtime=k8s

**Comment by caseydavenport** [comment]
Should be fixed in the v1.0.0-beta.  Adding to v1.0.0 milestone for re-test.

**Comment by djosborne** [comment]
Like @gunjan5 said, [this commit](https://github.com/projectcalico/calico-containers/commit/3a16a24d3821af9c0a4015d5153bb7088263768e) updates calicoctl to perform its status checks in the host namespace instead of bothering to exec on the container at all. 

I'm going to close, as I've reproed and this does seem to be resolved with calicoctl-v1.0.0-beta:
```
turk:~/ $ sudo calicoctl node run
Running command to load modules: modprobe -a xt_set ip6_tables
Enabling IPv4 forwarding
Enabling IPv6 forwarding
Increasing contrack limit
Running the following command:

docker run -d --net=host --privileged --name=calico-node -e CALICO_NETWORKING_BACKEND=bird -e NO_DEFAULT_POOLS= -e ETCD_AUTHORITY=127.0.0.1:2379 -e CALICO_LIBNETWORK_ENABLED=true -e IP= -e AS= -e ETCD_SCHEME=http -e ETCD_ENDPOINTS= -e HOSTNAME=durk -e IP6= -v /run/docker/plugins:/run/docker/plugins -v /var/run/docker.sock:/var/run/docker.sock -v /var/log/calico:/var/log/calico -v /var/run/calico:/var/run/calico -v /lib/modules:/lib/modules calico/node:v1.0.0-beta

turk:~/ $ sudo calicoctl node status
Calico process is running.

IPv4 BGP status
No IPv4 peers found.

IPv6 BGP status
No IPv6 peers found.

turk:~/ $ docker rename calico-node kitty-cat
turk:~/ $ docker ps
CONTAINER ID        IMAGE                     COMMAND             CREATED              STATUS              PORTS               NAMES
fa446a20a07e        calico/node:v1.0.0-beta   "start_runit"       About a minute ago   Up About a minute                       kitty-cat

turk:~/ $ sudo calicoctl node status                                             
Calico process is running.

IPv4 BGP status
No IPv4 peers found.

IPv6 BGP status
No IPv6 peers found.
```
